GVFS/GVFS.Virtualization/Projection/Readme.md

```diff
@@ -0,0 +1,181 @@
+
+```
+Example:
+
+For sparse folder entries of:
+foo/example
+other
+
+The SparseFolderData would be:
+root
+Children:
+|- foo (IsRecursive = false, Depth = 0)
+|  Children:
+|  |- example (IsRecursive = true, Depth = 1)
+|
+|- other (IsRecursive = true, Depth = 0)
+```
+
+### `GitIndexEntry`
+
+Class used to store the data from the index about a single entry.  There is only one instance of this class used during index parsing and it is reused for each index entry.  The reason for this is that version 4 of the git index has the [path prefix compressed](https://github.com/microsoft/git/blob/f5992bb185757a1654ce31424611b4d05bda3400/Documentation/technical/index-format.txt#L116) and the previous path is needed to create the path for the current entry.  The code in this class is heavily optimized to make parsing the index and the paths as fast as possible.
+
+### `GitIndexParser`
+
+Class that is responsible for parsing the git index based of version 4.  Please see [`index-format.txt`](https://github.com/microsoft/git/blob/f5992bb185757a1654ce31424611b4d05bda3400/Documentation/technical/index-format.txt) for detailed information of this format.  This is used to both validate the index and build the projection.  It currently ignores all index extensions and is only for getting the paths and building the tree using the `FolderData` and `FileData` classes. The index is read in chunks of 512K which gave the best performance.
+
+## Other classes
+
+### `ProjectedFileInfo`
+
+Class used to hold the data that is used by ProjFS when enumerating or creating placeholders.
+
+## `GitIndexProjection`
+
+Class used to hold the projection data and keep it up to date. This code uses and can be called from multiple threads.  It is using `ReaderWriterLockSlim` to synchronize access to the projection and ResetEvents for waiting and notification of events. There are caches for a variety of objects that are used.
+
+### Initialization
+
+Found in the `Initialize` method and does the following:
+
+1. Take a projection write lock
+2. Build the projection
+3. Release the write lock
+4. Update placeholders if needed
+5. Start the index parsing thread
+
+### Index parsing thread
+
+There is a thread started when the class is initialized that is waits to be woken up to parse the index. Events are used to indicate when the parsing is complete to make sure that the projection is in a good state before using it.
+
+When woken the parsing thread will:
+
+1. Check if it needs to stop
+2. Take the projection write lock
+3. Copy the index file and rebuild the projection while projection is invalid
+4. Release projection write lock
+5. If the projection was updated clear the negative path cache and update placeholders
+6. Set event indicating projection parsing is complete
+
+### Processes

```

Comment:
Consider moving this ahead of the internal classes section?  It's a good intro to the important high level flows, then the more detailed discussion of the class purposes can come later.